libcnb-test/src/lib.rs

```diff
@@ -85,15 +86,39 @@ impl IntegrationTest {
     /// - When the connection to Docker failed
     /// - When the internal Tokio runtime could not be created
     pub fn new(builder_name: impl Into<String>, app_dir: impl AsRef<Path>) -> Self {
+        let tokio_runtime =
+            tokio::runtime::Runtime::new().expect("Could not create internal Tokio runtime");
+
+        let docker = Docker::connect_with_ssl_defaults()
+            .and_then(|docker| {
+                // Bollard will not attempt to connect to Docker before an actual request. Because
+                // we want to ensure a connection can be established (see later comment) we need to
+                // ping Docker here.
+                match tokio_runtime.block_on(docker.ping()) {
+                    // If the connection via HTTPS failed and no explicit DOCKER_HOST has been
+                    // set, try to connect via a local socket or named pipe. This is a fallback
+                    // for local Docker Desktop configurations which currently do not expose
+                    // the HTTP API by default.
+                    Err(bollard::errors::Error::HyperResponseError { err })
+                        if err.is_connect()
+                            && env::var("DOCKER_HOST") == Err(VarError::NotPresent) =>
+                    {
+                        Docker::connect_with_local_defaults()
+                    }
+                    Ok(_) => Ok(docker), // Ping successful, use connection.
+                    Err(err) => Err(err), // Ping failed for other reasons than connection errors,
+                                          // keep error as-is for later display to user.
+                }
```

Comment:
As discussed on Slack just now, I think this won't work since Circle CI's Docker config needs SSL. When expanding the "Setup a remote Docker engine" section on [this CI run](https://app.circleci.com/pipelines/github/heroku/procfile-cnb-rust/114/workflows/c022ba84-945e-4d58-90e1-6eb1643531e7/jobs/229/steps), the env vars set are:

```
  DOCKER_CERT_PATH=/tmp/docker-certs1511074844
  DOCKER_HOST=tcp://ec2-3-92-79-176.compute-1.amazonaws.com:2376
  DOCKER_MACHINE_NAME=172854067
  DOCKER_TLS_VERIFY=1
  NO_PROXY=127.0.0.1,localhost,circleci-internal-outer-build-agent,ec2-3-92-79-176.compute-1.amazonaws.com:237
```
...which includes `DOCKER_TLS_VERIFY=1`.

Also, I think I would prefer to try and pick the best connection type up front, rather than potentially try the wrong option and then fallback every time (particularly since it will be invisible to the user/us, so we'd never know to fix it).

Ideally this would be part of bollard (eg `connect_with_defaults()`) - I think we should file an issue, but implement here in the meantime.

From searching around, eg:
https://github.com/search?q=org%3Amoby+DOCKER_HOST&type=Code
https://github.com/moby/moby/blob/3c06ebd876687555fdf030a3307a66908c4fa57c/client/client_unix.go
https://github.com/moby/moby/blob/3c06ebd876687555fdf030a3307a66908c4fa57c/client/client_windows.go
https://github.com/moby/moby/blob/3c06ebd876687555fdf030a3307a66908c4fa57c/client/client_test.go

...it seems that typical protocols for `DOCKER_HOST` are:
- `unix://`
- `npipe:////` (Windows)
- `tcp://`
- `http://` / `https://` (less common; presumably because of [this reason](https://github.com/docker/cli/blob/c9f84735128fa5fbf3ffa76b8940ea42aa675011/e2e/global/cli_test.go#L38-L39))

Now bollard's `connect_with_local_defaults()` (which on unix calls `connect_with_unix_defaults()`) already uses `DOCKER_HOST` iff it's set to `unix://`:
https://github.com/fussybeaver/bollard/blob/c5d87a4934c70a04f9c649fedb241dbd4943c927/src/docker.rs#L629-L641

Therefore given the above, and the fact that I don't think we should only support HTTPS (and not HTTP, since HTTPS is mandatory for Docker by default), I think here all we need to do here, is:

- if `DOCKER_HOST` set and starts with `tcp://` or `https://`, then use `connect_with_ssl_defaults()`
- otherwise use `connect_with_local_defaults()`

What do you think? :-)

Reply:
> What do you think?

I wrote 7d17cad before reading this... it implements it almost exactly like that :D I'll iterate a bit more and test on CircleCI - but I expect this to be the most viable route. :)

Reply:
> I think here all we need to do here, is:
> - If DOCKER_HOST set and starts with tcp:// or https://, then use connect_with_ssl_defaults() (this will mostly be used only in CI, or for users who have advanced configs locally)
> - otherwise use connect_with_local_defaults()

I was digging through pack's source, and it seems to match this pretty closely.